pkg/operator/controller/dns_status.go

```diff
@@ -92,34 +170,109 @@ func computeDNSDegradedCondition(oldCondition *operatorv1.OperatorCondition, clu
 		maxUnavailable, intstrErr := intstr.GetScaledValueFromIntOrPercent(&maxUnavailableIntStr, int(want), true)
 		switch {
 		case want == 0:
-			status = operatorv1.ConditionTrue
-			degradedReasons = append(degradedReasons, "NoDNSPodsDesired")
+			degradedConditions = append(degradedConditions, operatorv1.OperatorCondition{Type: DNSNoDNSPodsDesired, Status: operatorv1.ConditionTrue, LastTransitionTime: now})
 			messages = append(messages, "No DNS pods are desired; this could mean all nodes are tainted or unschedulable.")
 		case have == 0:
-			status = operatorv1.ConditionTrue
-			degradedReasons = append(degradedReasons, "NoDNSPodsAvailable")
+			degradedConditions = append(degradedConditions, operatorv1.OperatorCondition{Type: DNSNoDNSPodsAvailable, Status: operatorv1.ConditionTrue, LastTransitionTime: now})
 			messages = append(messages, "No DNS pods are available.")
 		case intstrErr != nil:
-			degradedReasons = append(degradedReasons, "InvalidDNSMaxUnavailable")
+			keepConditionUnknown = true
+			degradedConditions = append(degradedConditions, operatorv1.OperatorCondition{Type: DNSInvalidDNSMaxUnavailable, Status: operatorv1.ConditionUnknown, LastTransitionTime: now})
 			messages = append(messages, fmt.Sprintf("The DNS daemonset has an invalid MaxUnavailable value: %v", intstrErr))
 		case int(numberUnavailable) > maxUnavailable:
-			status = operatorv1.ConditionTrue
-			degradedReasons = append(degradedReasons, "MaxUnavailableDNSPodsExceeded")
+			degradedConditions = append(degradedConditions, operatorv1.OperatorCondition{Type: DNSMaxUnavailableDNSPodsExceeded, Status: operatorv1.ConditionTrue, LastTransitionTime: now})
 			messages = append(messages, fmt.Sprintf("Too many DNS pods are unavailable (%d > %d max unavailable).", numberUnavailable, maxUnavailable))
 		}
 	}
 
-	if len(degradedReasons) != 0 {
-		degradedCondition.Status = status
-		degradedCondition.Reason = strings.Join(degradedReasons, "")
-		degradedCondition.Message = strings.Join(messages, "\n")
-	} else {
-		degradedCondition.Status = operatorv1.ConditionFalse
-		degradedCondition.Reason = "AsExpected"
-		degradedCondition.Message = "Enough DNS pods are available, and the DNS service has a cluster IP address."
+	graceConditions, checkedConditions, requeueAfter := checkConditions(expectedConditions, degradedConditions)
+
+	// graceConditions have grace periods configured in the expectedConditions, so if present, this should be requeued
+	if len(graceConditions) != 0 {
+		finalDegradedCondition.Status = operatorv1.ConditionTrue
```

Comment:
The point of the grace period is that we report `Degraded=False` during the grace period, so shouldn't this be `finalDegradedCondition.Status = operatorv1.ConditionFalse`?  Also, as `checkConditions` has any conditions without a  grace period, shouldn't we check it first so that we do report `Degraded=True` immediately if there are any problems that don't have a grace period?

Reply:
You're right,  I prioritized the grace period evaluation and should not have, and should not mark it as Degraded during the grace period.  Fixed.